File: LeetcodeEasyProblems/_572. Subtree of Another Tree/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isSubtree(TreeNode root, TreeNode subRoot) {
        // If the subRoot are null, always return true
        if (subRoot == null){
            return true;
        }
        // if the root is empty while the subRoot is not, 
        // then return false
        if (root == null){
            return false;
        }
        // Now knowning that both of them are not empty
        // check if they are the same tree
        if (this.sameTree(root, subRoot)){
            return true;
        }
        // If they're not, then check the left and right subtree
        return this.isSubtree(root.left, subRoot)
        || this.isSubtree(root.right, subRoot);
        
    }

    public boolean sameTree(TreeNode p, TreeNode q){
        // If the root are both null, return true
        if (p == null && q == null){
            return true;
        }
        // If either of them are empty, return false
        if(p == null || q == null){
            return false;
        }
        // Now we know that they're both not empty
        // Check if their root.val are the same
        // If they're not, return false
        if (p.val != q.val){
            return false;
        }
        // Now we know that their root.val are the same
        // We can only return true when both their left 
        // and right subtree are the same
        return this.sameTree(p.left, q.left) && this.sameTree(p.right, q.right);
    }
}

File: LeetcodeEasyProblems/_572. Subtree of Another Tree/Solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def isSubtree(self, root, subRoot):
        """
        :type root: TreeNode
        :type subRoot: TreeNode
        :rtype: bool
        """
        # When the subroot is null, they are always 
        # the subtree of the root node tree
        if subRoot == None:
            return True
        # When the subroot is not null, while the root node is null
        # then return false
        if root == None:
            return False
        # Now after knowning that the root and subroot
        # are not null, check if they are the same tree
        if self.sameTree(root,subRoot):
            return True
        # Now branch out the left and right subtree of root node
        left_root = root.left
        right_root = root.right
        # And check if there's a subtree in them
        return self.isSubtree(left_root, subRoot) or self.isSubtree(right_root, subRoot)
    def sameTree(self, p, q):
        """
        :type p: TreeNode
        :type q: TreeNode
        :rtype: bool
        """
        # First we check if the root node is both null
        if p == None and q == None:
            return True
        # Check if either one of them is null
        if p == None or q == None:
            return False
        # Check if their values are different
        if p.val != q.val:
            return False
        return self.sameTree(p.right, q.right) and self.sameTree(p.left, q.left)
        